pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,959 @@
+
+            if total_collators == 0 {
+                return Zero::zero()
+            }
+
+            let reward_per_collator = current_distribution / BalanceOf::<T>::from(total_collators);
+
+            for (collator, _) in Collators::<T>::iter() {
+                Self::update_pending_claims(
+                    &collator,
+                    CircuitRole::Collator,
+                    reward_per_collator,
+                    BenefitSource::Inflation,
+                );
+            }
+
+            reward_per_collator.saturating_mul(BalanceOf::<T>::from(total_collators))
+        }
+
+        pub fn distribute_executor_rewards(current_distribution: BalanceOf<T>) -> BalanceOf<T> {
+            let accumulated_settlements = AccumulatedSettlements::<T>::iter().collect::<Vec<_>>();
+
+            // Get the total settled executions this round
+            let total_settled_executions_this_round =
+                match Self::total_settled_executions_this_round(accumulated_settlements.clone()) {
+                    Ok(total_settled_executions_this_round) => total_settled_executions_this_round,
+                    Err(e) => {
+                        log::error!(
+                            "Arithmetic Overflow when calculating settled executor rewards: {:?}",
+                            e
+                        );
+                        return Zero::zero()
+                    },
+                };
+
+            // Calculate the proportions of the total settled executions for each executor
+            let executions_proportionally_of_total_this_round =
+                Self::executions_proportionally_of_total_this_round(
+                    accumulated_settlements,
+                    total_settled_executions_this_round,
+                );
+
+            let mut distibuted_rewards = Zero::zero();
+
+            // Distribute the executor rewards proportionally
+            for (executor, accumulated_settlement, proportion) in
+                executions_proportionally_of_total_this_round
+            {
+                let reward = proportion.mul_ceil(current_distribution);
+
+                // Ensure the reward does not exceed 90% of the accumulated settlement amount
+                let max_reward = Perbill::from_percent(90).mul_ceil(accumulated_settlement);
+                let capped_reward = reward.min(max_reward);
+
+                // Update the pending claims for the executor
+                Self::update_pending_claims(
+                    &executor,
+                    CircuitRole::Executor,
+                    capped_reward,
+                    BenefitSource::Inflation,
+                );
+
+                // Remove the accumulated settlement from the storage
+                AccumulatedSettlements::<T>::remove(&executor);
+
+                distibuted_rewards += capped_reward;
+            }
+
+            distibuted_rewards
+        }
+
+        pub fn process_accumulated_settlements() -> Weight {
```

Comment:
Not really a fan of this, as it's inflexible, a complex call, and rewards incomplete. I see the benefit of having it claimable in an on-chain manner, but also a lot of downsides, when it comes to different things we might want to reward (tx value for all sfx, speed, liq on target, etc). Always requires us to add more complexity and strain on the node. Maybe we could think of a hybrid approach that allows us to make it claimable, but also more flexible

Reply:
process_accumulated_settlements is only adding the already-earned rewards for execution. 

distribute_executor_rewards adds the extra rewards on top of the executions. I see the point that extending those with extra rewards ruleset will enforce a runtime upgrade. 

we also have the bootstrap pool alloc. to apply the extra off-chain rules for distribution. I am up for keeping the simple subsidy of settlement_amount for executor for inflation as algorithmic rule at start